### PR TITLE
adding evaluation

### PR DESCRIPTION
**Pull request by liususan091219**
proposing to add an evaluation function so the user can see the evaluation metric after retraining.

Currently after retraining the metric is not observable. However, if we use estimator.model.evaluate to evaluate, it's also inconvenient for the TransformersEstimator, because transformer's model.evaluate takes the tokenized Dataset as the input instead of the raw X_val, y_val, which means we have to again tokenize X_val, y_val in the user code. This is against the design for minimizing user code, so I think it's necessary to add this function.

### CODE REVIEW

**Comment by liususan091219** [comment]
> > proposing to add an evaluation function so the user can see the evaluation metric after retraining.
> > Currently after retraining the metric is not observable. However, if we use estimator.model.evaluate to evaluate, it's also inconvenient for the TransformersEstimator, because transformer's model.evaluate takes the tokenized Dataset as the input instead of the raw X_val, y_val, which means we have to again tokenize X_val, y_val in the user code. This is against the design for minimizing user code, so I think it's necessary to add this function.
> 
> Could you clarify whether it's only for TransformerEstimator or for all estimators? What is the API you are proposing exactly?

I only need it to work for TransformerEstimator. I can change the code so it only works for TransformersEstimator. The api is (in test/nlp/test_autohf.py):

automl.evaluate(X_val, y_val, **{"metric": "accuracy"})

**Comment by sonichi** [comment]
> > > proposing to add an evaluation function so the user can see the evaluation metric after retraining.
> > > Currently after retraining the metric is not observable. However, if we use estimator.model.evaluate to evaluate, it's also inconvenient for the TransformersEstimator, because transformer's model.evaluate takes the tokenized Dataset as the input instead of the raw X_val, y_val, which means we have to again tokenize X_val, y_val in the user code. This is against the design for minimizing user code, so I think it's necessary to add this function.
> > 
> > 
> > Could you clarify whether it's only for TransformerEstimator or for all estimators? What is the API you are proposing exactly?
> 
> I only need it to work for TransformerEstimator. I can change the code so it only works for TransformersEstimator. The api is (in test/nlp/test_autohf.py):
> 
> automl.evaluate(X_val, y_val, **{"metric": "accuracy"})

How are we supposed to document this new method `AutoML.evaluate()`, if it is only valid for TransformerEstimator?
There is a related issue #466 .
My suggestion is, if you are going to add a method for the AutoML class, better make it work for all the estimators.
Otherwise, add a method for the TransformerEstimator class only.

File: flaml/automl.py [hunk 1]
```diff
@@ -246,7 +246,6 @@ def _compute_with_config_base(config_w_resource, state, estimator):
             * sample_size
             / state.data_size[0]
         )
-        # raise Exception("bbbbb", state.time_budget, budget)
 
         if _is_nlp_task(state.task):
```
Comment (sonichi) [inline flaml/automl.py:250]: Should this be moved to ml.py and for TransformerEstimator only?

File: flaml/automl.py [hunk 2]
```diff
@@ -344,7 +343,7 @@ def _train_with_config(
             estimator_class=self.learner_classes.get(estimator),
             budget=budget,
             fit_kwargs=self.fit_kwargs,
-            eval_metric="train_time",
+            eval_metric=self.metric if (hasattr(self, "metric")) else "train_time",
```
Comment (sonichi) [inline flaml/automl.py:346]:
```suggestion
            eval_metric=self.metric if hasattr(self, "metric") else "train_time",
```

File: flaml/model.py [hunk 3]
```diff
@@ -720,15 +719,11 @@ def _compute_metrics_by_dataset_name(self, eval_pred):
             metric_dict["automl_metric"] = loss
         return metric_dict
 
-    def _init_model_for_predict(self, X_test):
-        from datasets import Dataset
+    def _init_model_for_predict(self, test_dataset):
```
Comment (sonichi) [inline flaml/model.py:722]: Remove `test_dataset` from input and output. Remember to update all the references.

File: flaml/model.py [hunk 4]
```diff
@@ -753,22 +748,109 @@ def _init_model_for_predict(self, X_test):
...
 
-        new_trainer, test_dataset, _ = self._init_model_for_predict(X)
+        X_test, _ = self._preprocess(X, **self._kwargs)
+        test_dataset = Dataset.from_pandas(X_test)
+
+        new_trainer, test_dataset, _ = self._init_model_for_predict(test_dataset)
         predictions = new_trainer.predict(test_dataset)
         return predictions.predictions
 
+    def evaluate(self, X_val: DataFrame, y_val: Series, **kwargs):
```
Comment (sonichi) [inline flaml/model.py:765]: Is `score` a better name for this method? Shall we make the signature compatible with the `score()` In scikit-learn? See #466 .

File: flaml/nlp/utils.py [hunk 5]
```diff
@@ -13,6 +13,18 @@
     NLG_TASKS,
 )
 
+def load_default_checkpoint_path(task):
```
Comment (sonichi) [inline flaml/nlp/utils.py:16]: Where is this function used?

File: test/nlp/test_autohf.py [hunk 6]
```diff
@@ -128,6 +128,10 @@ def test_hf_data():
     )
 
     automl.predict_proba(X_test)
+
+    X_val = automl._preprocess(X_val)
+    automl._trained_estimator.evaluate(X_val, y_val, **{"metric": "accuracy"})
```
Comment (sonichi) [inline test/nlp/test_autohf.py:133]:
```suggestion
    automl.model.evaluate(X_val, y_val, **{"metric": "accuracy"})
```

File: test/nlp/test_autohf_custom_metric.py [hunk 7]
```diff
@@ -126,6 +126,8 @@ def test_custom_metric():
     automl.fit(
         X_train=X_train, y_train=y_train, X_val=X_val, y_val=y_val, **automl_settings
     )
+    X_val = automl._preprocess(X_val)
+    automl._trained_estimator.evaluate(X_val, y_val, **{"metric": custom_metric})
```
Comment (sonichi) [inline test/nlp/test_autohf_custom_metric.py:130]:
```suggestion
    automl.model.evaluate(X_val, y_val, **{"metric": custom_metric})
```

File: test/nlp/test_autohf.py [hunk 8]
```diff
@@ -128,6 +128,10 @@ def test_hf_data():
     )
 
     automl.predict_proba(X_test)
+
+    X_val = automl._preprocess(X_val)
```
Comment (sonichi) [inline test/nlp/test_autohf.py:132]: Is this intended for user's actual code or just for testing? For user's actual code, we shouldn't call `_preprocess()` as it is an internal function.
A followup question is what would be user's code? Need to add it to the example in documentation.

File: website/docs/Use-Cases/Task-Oriented-AutoML.md [hunk 9]
```diff
@@ -14,6 +14,7 @@
     - 'rank': learning to rank.
```
Comment (sonichi) [inline website/docs/Use-Cases/Task-Oriented-AutoML.md:14]:
```suggestion
    - 'ts_forecast_classification': time series forecasting for classification.
    - 'rank': learning to rank.
```

**Review by liususan091219 — COMMENTED** [review]
finished one iteration of review

**Review by liususan091219 — COMMENTED** [review]
reviewed

File: website/docs/Use-Cases/Task-Oriented-AutoML.md [hunk 10]
```diff
@@ -14,6 +14,7 @@
...
     - 'seq-classification': sequence classification.
     - 'seq-regression': sequence regression.
+    - 'multiplechoice-classification': multiple choice classification.
```
Comment (sonichi) [inline website/docs/Use-Cases/Task-Oriented-AutoML.md:17]: This change could conflict with my PR. Could you check? If it does, remove this change.

File: flaml/model.py [hunk 11]
```diff
@@ -234,6 +234,82 @@ class j.
...
+        from .ml import metric_loss_score
+
+        if self._model is not None:
+            if self._task == "rank":
+                raise NotImplementedError(
+                    "automl.score is not implemented for LGBMEstimator for ranking"
+                )
+            else:
+                X_val = self._preprocess(X_val)
+                self._metric = kwargs.get("metric", None)
```
[the remark this answers is not in the document]
Comment (liususan091219) [reply flaml/model.py:301]: Since this function .score is not in trainable function, it won't be pickled by ray, right?

File: flaml/model.py [hunk 12]
```diff
@@ -758,83 +834,29 @@ def predict_proba(self, X, **kwargs):
...
-            The evaluation score on the validation dataset.
-        """
+    def score(self, X_val: DataFrame, y_val: Series, **kwargs):
         import transformers
         from datasets import Dataset
 
         transformers.logging.set_verbosity_error()
 
-        self._metric = kwargs.get("metric", self._metric)
+        self._metric = kwargs["metric"]
```
[the remark this answers is not in the document]
Comment (liususan091219) [reply flaml/model.py:847]: I tested the custom metric case in test/nlp/test_autohf_custom_metric.py with ray, it does not raise the pickle error.

File: flaml/model.py [hunk 13]
```diff
@@ -1759,6 +1779,17 @@ def predict(self, X, **kwargs):
             )
             return np.ones(X.shape[0])
 
+    def score(self, X_val: DataFrame, y_val: Series, **kwargs):
+        from sklearn.metrics import r2_score
+        from .ml import metric_loss_score
+
+        y_pred = self.predict(X_val)
+        self._metric = kwargs.get("metric", None)
```
[the remark this answers is not in the document]
Comment (liususan091219) [reply flaml/model.py:1787]: Since this function .score is not in trainable function, it won't be pickled by ray, right?

File: test/automl/test_score.py [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+        try:
+            import prophet
+
+            automl.fit(
+                dataframe=df,
+                estimator_list=["prophet", "arima", "sarimax"],
+                **settings,
+                period=time_horizon,
+            )
+            automl.score(X_test, y_test)
```
[the remark this answers is not in the document]
Comment (liususan091219) [reply test/automl/test_score.py:44]: forecast is implemented, so if classification shouldn't raise NotImplementedError, same here

File: flaml/model.py [hunk 15]
```diff
@@ -88,7 +88,9 @@ def __init__(self, task="binary", **config):
 
         Args:
             task: A string of the task type, one of
-                'binary', 'multiclass', 'regression', 'rank', 'forecast'.
+                'binary', 'multiclass', 'regression', 'rank', 'forecast', 'seq-classification',
```
Comment (sonichi) [inline flaml/model.py:91]: Check my recent PR.
'ts_forecast', 'ts_forecast_classification'.

File: flaml/model.py [hunk 16]
```diff
@@ -293,22 +296,20 @@ def custom_metric(
 
         if self._model is not None:
             if self._task == "rank":
-                raise NotImplementedError(
-                    "automl.score is not implemented for LGBMEstimator for ranking"
-                )
+                raise NotImplementedError("automl.score is not implemented for ranking")
```
Comment (sonichi) [inline flaml/model.py:299]:
```suggestion
                raise NotImplementedError("AutoML.score() is not implemented for ranking")

```

File: flaml/model.py [hunk 17]
```diff
@@ -293,22 +296,20 @@ def custom_metric(
...
             else:
                 X_val = self._preprocess(X_val)
-                self._metric = kwargs.get("metric", None)
-                if self._metric:
+                metric = kwargs.get("metric", None)
+                if metric:
                     y_pred = self.predict(X_val, **kwargs)
-                    return metric_loss_score(self._metric, y_pred, y_val)
+                    return 1.0 - metric_loss_score(metric, y_pred, y_val)
```
Comment (sonichi) [inline flaml/model.py:305]: score is not always 1-loss.

File: website/docs/Use-Cases/Task-Oriented-AutoML.md [hunk 18]
```diff
@@ -14,7 +14,7 @@
     - 'rank': learning to rank.
     - 'seq-classification': sequence classification.
     - 'seq-regression': sequence regression.
-    - 'multiplechoice-classification': multiple choice classification.
+    - 'multiplechoice-classification': multipe choice classification.
```
Comment (sonichi) [inline website/docs/Use-Cases/Task-Oriented-AutoML.md:17]: still conflicting

File: flaml/ml.py [hunk 19]
```diff
@@ -219,6 +219,12 @@ def is_in_sklearn_metric_name_set(metric_name):
     return metric_name.startswith("ndcg") or metric_name in sklearn_metric_name_set
 
 
+def is_min_metric(metric_name):
+    if metric_name in ["rmse", "mae", "mse", "log_loss", "mape"]:
+        return True
+    return False
```
Comment (sonichi) [inline flaml/ml.py:225]:
```suggestion
    return metric_name in ["rmse", "mae", "mse", "log_loss", "mape"]
```

File: flaml/model.py [hunk 20]
```diff
@@ -88,7 +88,9 @@ def __init__(self, task="binary", **config):
 
         Args:
             task: A string of the task type, one of
-                'binary', 'multi', 'regression', 'rank', 'forecast'.
+                'binary', 'multiclass', 'regression', 'rank', 'seq-classification',
+                'seq-regression', 'token-classification', 'multichoice-classification', 'ts_forecast', 'ts_forecast_classification'.
+                'summarization', 'ts_forecast', 'ts_forecast_classification'.
```
Comment (sonichi) [inline flaml/model.py:93]: 'ts_forecast' and 'ts_forecast_classification' are repeated twice.

File: flaml/model.py [hunk 21]
```diff
@@ -250,59 +250,32 @@ def score(self, X_val: DataFrame, y_val: Series, **kwargs):
...
-                        trainer = estimator._trainer
-                    if y_test is not None:
-                        X_test, _ = estimator._preprocess(X_test)
-                        eval_dataset = Dataset.from_pandas(TransformersEstimator._join(X_test, y_test))
-                    else:
-                        X_test, _ = estimator._preprocess(X_test)
-                        eval_dataset = Dataset.from_pandas(X_test)
+                You can also pass a customized metric function, for examples on how to pass a
+                customized metric function, please check test/nlp/test_autohf_custom_metric.py and
+                test/automl/test_multiclass.py.
```
Comment (sonichi) [inline flaml/model.py:255]: add a hyperlink using []()

File: flaml/model.py [hunk 22]
```diff
@@ -250,59 +250,32 @@ def score(self, X_val: DataFrame, y_val: Series, **kwargs):
...
             else:
                 X_val = self._preprocess(X_val)
                 metric = kwargs.get("metric", None)
                 if metric:
                     y_pred = self.predict(X_val, **kwargs)
-                    return 1.0 - metric_loss_score(metric, y_pred, y_val)
+                    if is_min_metric(metric):
+                        return metric_loss_score(metric, y_pred, y_val)
+                    else:
+                        return 1.0 - metric_loss_score(metric, y_pred, y_val)
```
Comment (sonichi) [inline flaml/model.py:278]: are all the other maximization scores equal to 1-loss?

File: test/automl/test_score.py [hunk 23]
```diff
@@ -201,4 +201,4 @@ def test_rank(self):
 
 if __name__ == "__main__":
     test = TestClassification()
-    test.test_classification()
+    test.test_regression()
```
Comment (sonichi) [inline test/automl/test_score.py:204]: TestClassification should be renamed.

File: flaml/model.py (hunk 22 above)
Comment (liususan091219) [reply flaml/model.py:279]: updated is_min_metric

File: flaml/model.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/model.py:822]: Please try explicitly pickle the object after .score().

File: flaml/model.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/model.py:1762]: Please try explicitly pickle the object after .score().

**Review by sonichi — COMMENTED** [review]
Please see my reply.

File: flaml/model.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/model.py]: Please try explicitly pickle after .score().

File: flaml/model.py [hunk 24]
```diff
@@ -234,6 +236,55 @@ class j.
         X = self._preprocess(X)
         return self._model.predict_proba(X)
 
+    def score(self, X_val: DataFrame, y_val: Series, **kwargs):
+        """Report the evaluation score of BaseEstimator.
```
Comment (sonichi) [inline flaml/model.py:240]:
```suggestion
        """Report the evaluation score of a trained estimator.

```

**Review by sonichi — APPROVED** [review]
